### ISSUE TITLE
Demonstrate parallel execution of a loss function

### ISSUE THREAD

**Issue by jackbaker1001** [XanaduAI/GradDFT#50]
On an HPC cluster, each term in a mean square loss can be calculated using embarrassingly parallel logic.

Unfortunately, the native way of doing this with `jax` (using `jax.vmap` and `jax.pmap`) is not compatible with input we must parallelize over: the `Molecule` object. This is because its data is stored in "ragged" structure. I.e., the dimensions of the grid for one molecule are very often different from the grid for another and the dimensions of the 1-RDM for one molecule is different for another: `jnp.array([rdm1_1, rdm1_2])` will not work.

This means that for loss parallelism, we need to think differently. Sharding may be the way forward, but this requires more thought. A good reference is here: https://jax.readthedocs.io/en/latest/notebooks/Distributed_arrays_and_automatic_parallelization.html

I don't think we will get around to solving this problem before our release deadline, but if we want to do something with HPC, getting this right is non-negotiable.

**Comment by jackbaker1001** [comment]
Related to #83

**Comment by jackbaker1001** [comment]
Having playing around with the [multiple hosts parallelism](https://jax.readthedocs.io/en/latest/multi_process.html) in JAX, I came across many issues on Perlmutter with the detection of GPUs. 

I'm giving [mpi4jax](https://github.com/mpi4jax/mpi4jax) a go for this task now. It should be fairly easy if this works well on Perlmutter.
